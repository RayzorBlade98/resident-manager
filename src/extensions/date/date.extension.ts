declare interface Date {
  /**
   * Converts the date to a new date in UTC timezone
   */
  toUTC(): Date;

  /**
   * Converts the date to 'DD.MM.YYYY' format string
   */
  toPreferredString(): string;
}

function toUTC(this: Date): Date {
  return new Date(
    Date.UTC(this.getFullYear(), this.getMonth(), this.getDate(), 0, 0, 0, 0),
  );
}
Date.prototype.toUTC = toUTC;

function toPreferredString(this: Date): string {
  const day: number = this.getDate();
  const month: number = this.getMonth() + 1;
  const year = String(this.getFullYear());

  const dayPrefix: string = day < 10 ? '0' : '';
  const monthPrefix: string = month < 10 ? '0' : '';

  return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
}
Date.prototype.toPreferredString = toPreferredString;
