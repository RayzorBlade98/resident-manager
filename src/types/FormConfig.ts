import { GroupedFormProps } from '_/components/form/GroupedForm/GroupedForm';
import { FormValidationArguments } from '_/hooks/useFormValidation/useFormValidation';

/**
 * Object containing the config of a grouped form
 */
export type FormConfig<TInput extends object, TGroups extends string> = {
  formValidationConfig: Omit<
  FormValidationArguments<TInput>,
  'onSubmitSuccess'
  >;
  formGroupConfig: Omit<
  GroupedFormProps<TInput, TGroups>,
  'children' | 'formErrors'
  >;
};
