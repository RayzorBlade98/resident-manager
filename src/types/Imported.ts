import Prettify from './Prettify';

/**
 * Type that replaces all date types with the string type
 * representing the result of the import of an object from type `T`
 */
type Imported<T extends object> = Prettify<{
  [key in keyof T]: T[key] extends Date | undefined
    ? string
    : T[key] extends object
      ? Imported<T[key]>
      : T[key];
}>;

export default Imported;
