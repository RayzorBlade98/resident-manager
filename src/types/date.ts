/**
 * Enum containing all months as `string`
 */
export enum Month {
  January = 'Januar',
  Febuary = 'Februar',
  March = 'MÃ¤rz',
  April = 'April',
  May = 'Mai',
  June = 'Juni',
  July = 'Juli',
  August = 'August',
  September = 'September',
  October = 'Oktober',
  November = 'November',
  December = 'Dezember',
}

/**
 * Type representing a year
 */
export type Year = number;

/**
 * Information about a specific month and year
 */
export interface MonthYear {
  month: Month;
  year: Year;
}

/**
 * Type of the string representation of a `MonthYear`
 *
 * Format: "[month] [year]"
 */
export type MonthYearString = `${Month} ${Year}`;

/**
 * Type of the string representation of a date including day, month and year
 *
 * Format: `dd.mm.yyyy`
 */
export type DateString = `${string}.${string}.${string}`;

/**
 * Type of the date string representation returned by the date input field
 *
 * Format: `yyyy-mm-dd`
 */
export type DateInputString = `${string}-${string}-${string}`;

/**
 * Class that provides utility functions for `MonthYear` objects
 */
export abstract class MonthYearUtils {
  /**
   * Adds the specified number of months to the given `MonthYear` object
   * @param monthYear `MonthYear` object to which the months should be added
   * @param addedMonths Number of months that should be added (can be negative)
   * @returns new `MonthYear` object with the adjusted month (and year)
   */
  public static addMonths(monthYear: MonthYear, addedMonths = 1): MonthYear {
    if (addedMonths < 0) {
      // Add negative amount of months == subtract months
      return MonthYearUtils.subtractMonths(monthYear, addedMonths * -1);
    }

    const newMonthYear = { ...monthYear };
    if (newMonthYear.month === Month.December) {
      // Year change
      newMonthYear.month = Month.January;
      newMonthYear.year += 1;
    } else {
      // Add one month to the object
      const months = Object.values(Month);
      const oldMonthIndex = months.findIndex(
        (m: Month) => m === newMonthYear.month,
      );
      newMonthYear.month = months[oldMonthIndex + 1];
    }

    // Call `addMonths` recursive until all months got addded
    if (addedMonths !== 1) {
      return MonthYearUtils.addMonths(newMonthYear, addedMonths - 1);
    }
    return newMonthYear;
  }

  /**
   * Checks if two `MonthYear` objects are equal
   * @param a first `MonthYear` object to compare
   * @param b second `MonthYear` object to compare
   * @returns `true` if both the month and year are the same, else `false`
   */
  public static areEqual(a: MonthYear, b: MonthYear): boolean {
    return a.month === b.month && a.year === b.year;
  }

  /**
   * Compares two `MonthYear` objects
   * @param a First `MonthYear` object to compare
   * @param b Second `MonthYear` object to compare
   * @returns
   *
   * - 0 if both object have the same month and year
   * - <0 if the first object contains a date that's earlier than the second object
   * - \>0 if the first object contains a date that's later than the second object
   */
  public static compare(a: MonthYear, b: MonthYear): number {
    if (a.year !== b.year) return a.year - b.year;

    const months = Object.values(Month);
    const aMonthIndex = months.findIndex((m: Month) => m === a.month);
    const bMonthIndex = months.findIndex((m: Month) => m === b.month);
    return aMonthIndex - bMonthIndex;
  }

  /**
   * Returns a `MonthYear` object that contains information of the current month and year
   * @returns `MonthYear` object for the current date
   */
  public static getCurrentMonthYear(): MonthYear {
    const date: Date = new Date();
    const monthIndex: number = date.getUTCMonth();
    const month: Month = Object.values(Month)[monthIndex] as Month;
    const year: Year = date.getUTCFullYear();

    return {
      month,
      year,
    };
  }

  /**
   * Parses a string representation of a `MonthYear` object to its respective object
   * @param monthYear string representation of a `MonthYear` object that should be parsed
   * @returns `MonthYear` object that contains the information of the specified `string`
   */
  public static parseString(monthYear: MonthYearString): MonthYear {
    const splits = monthYear
      .split(/(\s+)/)
      .filter((s: string) => s.trim().length > 0);
    return {
      month: splits[0] as Month,
      year: Number(splits[1]),
    };
  }

  /**
   * Subtracts the specified number of months from the given `MonthYear` object
   * @param monthYear `MonthYear` object from which the months should be subtraced
   * @param subtractedMonths Number of months that should be subtracted (can be negative)
   * @returns new `MonthYear` object with the adjusted month (and year)
   */
  public static subtractMonths(
    monthYear: MonthYear,
    subtractedMonths = 1,
  ): MonthYear {
    if (subtractedMonths < 0) {
      // Subtract negative amount of months == add months
      return MonthYearUtils.addMonths(monthYear, subtractedMonths * -1);
    }

    const newMonthYear = { ...monthYear };
    if (newMonthYear.month === Month.January) {
      // Year change
      newMonthYear.month = Month.December;
      newMonthYear.year -= 1;
    } else {
      // Subtract one month from the object
      const months = Object.values(Month);
      const oldMonthIndex = months.findIndex(
        (m: Month) => m === newMonthYear.month,
      );
      newMonthYear.month = months[oldMonthIndex - 1];
    }

    // Call `subtractMonths` recursive until all months got subtracted
    if (subtractedMonths !== 1) {
      return MonthYearUtils.subtractMonths(newMonthYear, subtractedMonths - 1);
    }
    return newMonthYear;
  }

  /**
   * Converts a `MonthYear` to its string representation
   * @param monthYear `MonthYear` that should be converted
   * @returns string representation of the specified `MonthYear` object with the format "[month] [year]"
   */
  public static toString(monthYear: MonthYear): MonthYearString {
    return `${monthYear.month} ${monthYear.year}`;
  }

  /**
   * Creates a list of `MonthYear` objects that contains each month
   * between the two specified `MonthYear` objects (inclusive these months)
   * @param start Start of the timespan
   * @param end End of the timespan
   * @returns List of `MonthYear` objects with all months between `start` and `end`
   */
  public static timespan(start: MonthYear, end: MonthYear): MonthYear[] {
    const timespan = [{ ...start }];

    // Whether to add a month each loop or subtract one
    const direction = Math.sign(MonthYearUtils.compare(end, start));

    if (MonthYearUtils.areEqual(start, end)) return timespan;

    for (
      let next = MonthYearUtils.addMonths(start, direction);
      !MonthYearUtils.areEqual(next, end);
      next = MonthYearUtils.addMonths(next, direction)
    ) {
      timespan.push(next);
    }
    timespan.push({ ...end });
    return timespan;
  }
}

/**
 * Class that provides utility functions for dates containing day, month and year.
 */
export abstract class DateUtils {
  /**
   * Converts a date string from `yyyy-mm-dd` to `dd.mm.yyyy` format
   * @param date Date in `yyyy-mm-dd` format
   * @returns Converted date in `dd.mm.yyyy` format
   */
  public static convertDateInputString(date: DateInputString): DateString {
    const splits = date.split('-');
    return `${splits[2]}.${splits[1]}.${splits[0]}`;
  }

  /**
   * Converts a date string from `dd.mm.yyyy` to `yyyy-mm-dd` format
   * @param date Date in `dd.mm.yyyy` format
   * @returns Converted date in `yyyy-mm-dd` format
   */
  public static convertToDateInputString(date: DateString): DateInputString {
    const splits = date.split('.');
    return `${splits[2]}-${splits[1]}-${splits[0]}`;
  }

  /**
   * Returns a date string containing the current date
   * @returns Date string containing the current date in `dd.mm.yyyy` format
   */
  public static getCurrentDate(): DateString {
    const date: Date = new Date();
    const day: number = date.getUTCDate();
    const month: number = date.getUTCMonth() + 1;
    const year = String(date.getUTCFullYear());

    const dayPrefix: string = day < 10 ? '0' : '';
    const monthPrefix: string = month < 10 ? '0' : '';

    return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
  }
}
