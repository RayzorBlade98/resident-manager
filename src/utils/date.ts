// Remove after adding another export:
/* eslint-disable import/prefer-default-export */

/**
 * Converts the given date to its string representaion (format: DD.MM.YYYY)
 * @param date date that should be converted
 * @returns string representation of the date
 */
export function dateToString(date: Date): string {
  const day: number = date.getDate();
  const month: number = date.getMonth() + 1;
  const year = String(date.getFullYear());

  const dayPrefix: string = day < 10 ? '0' : '';
  const monthPrefix: string = month < 10 ? '0' : '';

  return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
}
