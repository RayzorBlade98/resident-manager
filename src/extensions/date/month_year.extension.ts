import './date.extension';

const MONTH_NAMES = [
  'Januar',
  'Februar',
  'MÃ¤rz',
  'April',
  'Mai',
  'Juni',
  'Juli',
  'August',
  'September',
  'Oktober',
  'November',
  'Dezember',
];

/**
 * Date object that represents a month and a year
 */
class MonthYear extends Date {
  /**
   * Creates a new `MonthYear` object
   * @param month month of the new object. If undefined the current month will be used.
   * @param year year of the new object. If undefined the current year will be used.
   */
  constructor(month?: number, year?: number) {
    const now = new Date();
    const utc = new Date(
      year !== undefined ? year : now.getFullYear(),
      month !== undefined ? month : now.getMonth(),
      1,
    ).toUTC();
    super(utc.getTime());
  }

  /**
   * Converts a `Date` object to a `MonthYear`
   * @param date Date that should be converted
   * @returns `MonthYear` with the same month and year as `date`
   */
  public static fromDate(date: Date): MonthYear {
    return new MonthYear(date.getMonth(), date.getFullYear());
  }

  /**
   * Converts a date string to a `MonthYear`
   * @param date Date string that should be converted
   * @returns `MonthYear` with the same month and year as `date`
   */
  public static fromString(date: string): MonthYear {
    return MonthYear.fromDate(new Date(date));
  }

  /**
   * Adds the specified number of months to the given `MonthYear` object
   * @param toAdd Number of months that should be added (can be negative)
   * @returns this month year
   */
  public addMonths(toAdd = 1): MonthYear {
    this.setMonth(this.getMonth() + toAdd);
    return this;
  }

  /**
   * Creates a copy this object.
   * @returns the copy with the same month and year
   */
  public clone(): MonthYear {
    return new MonthYear(this.getMonth(), this.getFullYear());
  }

  /**
   * Checks whether this object has the same month and year as anoter `MonthYear`
   * @param other true if both the month and year are the same, else false
   */
  public equals(other: MonthYear): boolean {
    return this >= other && this <= other;
  }

  /**
   * Creates a list of `MonthYear` objects that contains each month
   * between the two specified `MonthYear` objects (inclusive these months)
   * @param start Start of the timespan
   * @param end End of the timespan
   * @returns List of `MonthYear` objects with all months between `start` and `end`
   */
  public static timespan(start: MonthYear, end: MonthYear): MonthYear[] {
    // Whether to add a month each loop or subtract one
    const direction = end > start ? 1 : -1;

    const timespan: MonthYear[] = [start.clone()];
    let next = start;
    while (!end.equals(next)) {
      next = next.clone();
      next.addMonths(direction);
      timespan.push(next);
    }

    return timespan;
  }

  /**
   * Generates a string representation in the format `<month name> <year>`
   * @returns string representation of this object
   */
  public toString(): string {
    return `${MONTH_NAMES[this.getMonth()]} ${this.getFullYear()}`;
  }
}

export default MonthYear;
