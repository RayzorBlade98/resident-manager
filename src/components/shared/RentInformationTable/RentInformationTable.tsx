import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from '@mui/material';
import React, { useMemo } from 'react';
import { convertCurrencyCentsToString } from '../../../utils/currency/currency.utils';
import RentInformationUtils from '../../../utils/rent/rent.utils';
import AddRentPaymentModal from './AddRentPaymentModal/AddRentPaymentModal';
import PaymentStatusIcon from './PaymentStatusIcon/PaymentStatusIcon';
import { AddPaymentIcon } from '_/components/generic/ModalIconButton/AddPaymentIcon/AddPaymentIcon';
import { OpenDocumentButton } from '_/components/generic/buttons/OpenDocumentButton/OpenDocumentButton';
import MonthYear from '_/extensions/date/month_year.extension';
import { PaymentStatus } from '_/models/resident/rent';
import { Resident } from '_/models/resident/resident';

interface RentInformationTableProps {
  /**
   * Resident for which the rent information should be displayed
   */
  resident: Resident;

  /**
   * First month for which the rent information should be displayed (optional)
   */
  start?: MonthYear;

  /**
   * Last month for which the rent information should be displayed (optional)
   */
  end?: MonthYear;
}

/**
 * Table that displays all rent informations
 */
function RentInformationTable(props: RentInformationTableProps): JSX.Element {
  const filteredRentInformation = useMemo(
    () => props.resident.rentInformation.filter(
      (i) => (!props.start || i.dueDate >= props.start)
          && (!props.end || i.dueDate <= props.end),
    ),
    [props],
  );

  return (
    <TableContainer sx={{ height: '90%' }}>
      <Table stickyHeader>
        <TableHead>
          <TableRow>
            <TableCell>Monat</TableCell>
            <TableCell>Miete</TableCell>
            <TableCell>Nebenkosten</TableCell>
            <TableCell>Bezahlt</TableCell>
            <TableCell>Überweisung</TableCell>
            <TableCell>Aktionen</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {filteredRentInformation.map((rent) => (
            <TableRow
              key={`${props.resident.id}-${rent.dueDate.toString()}`}
              sx={{
                height: '73px',
                '&:last-child td, &:last-child th': { border: 0 },
              }}
            >
              <TableCell>{rent.dueDate.toString()}</TableCell>
              <TableCell>{convertCurrencyCentsToString(rent.rent)}</TableCell>
              <TableCell>
                {convertCurrencyCentsToString(rent.incidentals)}
              </TableCell>
              <TableCell>
                <PaymentStatusIcon rentInformation={rent} />
              </TableCell>
              <TableCell>
                <OpenDocumentButton
                  documentId={rent.bankTransferDocumentId}
                  documentTarget={{
                    type: 'resident',
                    residentId: props.resident.id,
                  }}
                  tooltip="Überweisung anzeigen"
                />
              </TableCell>
              <TableCell>
                <AddPaymentIcon
                  modal={(modalProps) => (
                    <AddRentPaymentModal
                      {...modalProps}
                      resident={props.resident}
                      rentInformation={rent}
                    />
                  )}
                  hidden={
                    RentInformationUtils.getPaymentStatus(rent)
                    !== PaymentStatus.Unpaid
                  }
                />
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
}

export default RentInformationTable;
