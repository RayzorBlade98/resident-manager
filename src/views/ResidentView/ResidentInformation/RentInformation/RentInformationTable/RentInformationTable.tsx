import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from '@mui/material';
import React from 'react';
import { useRecoilValue } from 'recoil';
import { convertCurrencyCentsToString } from '../../../../../utils/currency/currency.utils';
import AddPaymentIcon from './AddPaymentIcon/AddPaymentIcon';
import PaymentStatusIcon from './PaymentStatusIcon/PaymentStatusIcon';
import { residentViewSelectedResidentState } from '_/views/ResidentView/states/resident_view_state';

/**
 * Table that displays all rent informations
 */
function RentInformationTable(): JSX.Element {
  const rentInformation = useRecoilValue(
    residentViewSelectedResidentState,
  )?.rent;

  return (
    <TableContainer>
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>Monat</TableCell>
            <TableCell>Miete</TableCell>
            <TableCell>Nebenkosten</TableCell>
            <TableCell>Bezahlt</TableCell>
            <TableCell>Aktionen</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {rentInformation?.map((rent) => (
            <TableRow
              key={rent.dueDate.toString()}
              sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
            >
              <TableCell>{rent.dueDate.toString()}</TableCell>
              <TableCell>{convertCurrencyCentsToString(rent.rent)}</TableCell>
              <TableCell>
                {convertCurrencyCentsToString(rent.incidentals)}
              </TableCell>
              <TableCell>
                <PaymentStatusIcon rentInformation={rent} />
              </TableCell>
              <TableCell>
                <AddPaymentIcon rentInformation={rent} />
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
}

export default RentInformationTable;
