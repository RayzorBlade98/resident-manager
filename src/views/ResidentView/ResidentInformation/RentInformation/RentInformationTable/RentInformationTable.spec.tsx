import { render } from '@testing-library/react';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { setRecoil } from 'recoil-nexus';
import RentInformationTable from './RentInformationTable';
import residentState from '_/states/saveStates/resident_state';
import { Month } from '_/types/date';
import residentViewState from '_/views/ResidentView/states/resident_view_state';
import RecoilTestWrapper from '_tests/__test_utils__/RecoillTestWrapper';
import RentInformationBuilder from '_tests/__test_utils__/builders/rent_information_builder';
import ResidentBuilder from '_tests/__test_utils__/builders/resident_builder';

describe('RentInformationTable', () => {
  test('should match snapshot', () => {
    // Arrange
    const resident = new ResidentBuilder()
      .addRentInformation(
        new RentInformationBuilder()
          .withRent(10)
          .withIncidentals(5)
          .withMonth(Month.May)
          .build(),
      )
      .addRentInformation(
        new RentInformationBuilder()
          .withRent(20)
          .withIncidentals(15)
          .withMonth(Month.April)
          .withPayment(20, new Date(2023, 3, 1))
          .build(),
      )
      .addRentInformation(
        new RentInformationBuilder()
          .withRent(30)
          .withIncidentals(25)
          .withMonth(Month.March)
          .withPayment(55, new Date(2023, 2, 1))
          .build(),
      )
      .build();

    const renderResult = render(
      <RecoilTestWrapper>
        <RentInformationTable />
      </RecoilTestWrapper>,
    );

    act(() => {
      setRecoil(residentState, [resident]);
      setRecoil(residentViewState, (state) => ({
        ...state,
        selectedResident: resident.id,
      }));
    });

    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });
});
