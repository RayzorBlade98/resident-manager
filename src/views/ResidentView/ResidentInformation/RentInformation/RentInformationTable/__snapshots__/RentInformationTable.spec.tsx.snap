// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RentInformationTable should match snapshot 1`] = `
<div>
  <div
    class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
  >
    <table
      class="MuiTable-root css-rqglhn-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Monat
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Miete
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Nebenkosten
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Bezahlt
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Aktionen
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-34nofg-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Mai 2023
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            0.10€
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            0.05€
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <svg
              aria-hidden="true"
              aria-label="Unbezahlt"
              class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium css-1wxstni-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="HighlightOffIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M14.59 8 12 10.59 9.41 8 8 9.41 10.59 12 8 14.59 9.41 16 12 13.41 14.59 16 16 14.59 13.41 12 16 9.41 14.59 8zM12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <svg
              aria-hidden="true"
              aria-label="Zahlung hinzufügen"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-tpuggo-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="PaymentsOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 14V6c0-1.1-.9-2-2-2H3c-1.1 0-2 .9-2 2v8c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2zm-2 0H3V6h14v8zm-7-7c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3zm13 0v11c0 1.1-.9 2-2 2H4v-2h17V7h2z"
              />
            </svg>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-34nofg-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            April 2023
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            0.20€
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            0.15€
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <svg
              aria-hidden="true"
              aria-label="Teilweise bezahlt am 01.04.2023 (0.20€ von 0.35€)"
              class="MuiSvgIcon-root MuiSvgIcon-colorWarning MuiSvgIcon-fontSizeMedium css-3adcp6-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="CheckCircleOutlineIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 7.58 10 14.17l-3.59-3.58L5 12l5 5 8-8zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <svg
              aria-hidden="true"
              aria-label="Zahlung hinzufügen"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-tpuggo-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="PaymentsOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 14V6c0-1.1-.9-2-2-2H3c-1.1 0-2 .9-2 2v8c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2zm-2 0H3V6h14v8zm-7-7c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3zm13 0v11c0 1.1-.9 2-2 2H4v-2h17V7h2z"
              />
            </svg>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-34nofg-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            März 2023
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            0.30€
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            0.25€
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <svg
              aria-hidden="true"
              aria-label="Bezahlt am 01.03.2023"
              class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeMedium css-4b67si-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="CheckCircleOutlineIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 7.58 10 14.17l-3.59-3.58L5 12l5 5 8-8zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <svg
              aria-hidden="true"
              aria-label="Zahlung hinzufügen"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-tpuggo-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="PaymentsOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 14V6c0-1.1-.9-2-2-2H3c-1.1 0-2 .9-2 2v8c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2zm-2 0H3V6h14v8zm-7-7c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3zm13 0v11c0 1.1-.9 2-2 2H4v-2h17V7h2z"
              />
            </svg>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
